## Arindra Nurshadrina Ramadini
## 2209106070 / B2


## Posttest 5 - Deep Learning

## Import Library

In [1]:
# import library yang dibutuhkan

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical


## Load dan Read Dataset (ambil dari library)

In [2]:
# Load dataset adult income
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
column_names = [
    "age", "workclass", "fnlwgt", "education", "education-num",
    "marital-status", "occupation", "relationship", "race", "sex",
    "capital-gain", "capital-loss", "hours-per-week", "native-country", "income"
]


In [3]:
# read dataset

df = pd.read_csv(url, header=None, names=column_names, na_values=' ?', skipinitialspace=True)


Merupakan dataset Adult-Income, yang mengklasifikasikan income <50k dan >50k

## Informasi Dataset

In [11]:
# cek informasi dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  int64 
dtypes: int64(7), object(8)
memory usage: 3.7+ MB


In [12]:
# cek apakah ada nilai null
print(df.isnull().sum())


age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64


Tidak ada null value

## Convert dan Define (feature, target)

In [14]:
# Convert variabel target ke binary (0: <=50K, 1: >50K)
df['income'] = np.where(df['income'] == '>50K', 1, 0)

# definisi feature dan target
X = df.drop('income', axis=1)
y = df['income']


## Encoding Categorical Feature

In [13]:
# One-hot encode feature kategorikal
X = pd.get_dummies(X, drop_first=True)

## Splitting

In [5]:
# Split dataset training (80%) and validation (20%)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)


## Standarisasi

In [6]:
# Standardize features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Convert label target ke categorical
y_train = to_categorical(y_train)
y_val = to_categorical(y_val)



*   Standarisasi agar semua feature memiliki skala yang sama
*   Lalu selanjutnya mengonversi label target ke dalam bentuk one-hot encoding



In [7]:
# membuat model sequential
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(2, activation='softmax'))  # Output layer untuk binary classification


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [8]:
# melihat bentukan model yg dibuat

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 64)                  │           6,464 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 2)                   │              66 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 8,610 (33.63 KB)

 Trainable params: 8,610 (33.63 KB)

 Non-trainable params: 0 (0.00 B)

## Model Training

In [9]:
# Compile model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train model
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, batch_size=32)


Epoch 1/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8035 - loss: 0.4062 - val_accuracy: 0.8475 - val_loss: 0.3293
Epoch 2/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8517 - loss: 0.3183 - val_accuracy: 0.8549 - val_loss: 0.3184
Epoch 3/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8551 - loss: 0.3113 - val_accuracy: 0.8526 - val_loss: 0.3172
Epoch 4/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8599 - loss: 0.2983 - val_accuracy: 0.8567 - val_loss: 0.3155
Epoch 5/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8647 - loss: 0.2936 - val_accuracy: 0.8600 - val_loss: 0.3164
Epoch 6/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8639 - loss: 0.2880 - val_accuracy: 0.8538 - val_loss: 0.3184
Epoch 7/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8675 - loss: 0.2857 - val_accuracy: 0.8531 - val_loss: 0.3226
Epoch 8/20
814/814 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8726 - loss: 0.2776 - val_accuracy: 0.

menggunakan optimizer adam, lalu fungsi loss menggunakan categorical_crossentropy karena target dikonversi menjadi bentuk one-hot encoding.

## Evaluate Model

In [10]:
# evaluasi model
train_loss, train_acc = model.evaluate(X_train, y_train, verbose=0)

val_loss, val_acc = model.evaluate(X_val, y_val, verbose=0)

# Print akurasi
print(f"Training Accuracy: {train_acc * 100:.2f}%")
print(f"Validation Accuracy: {val_acc * 100:.2f}%")


Training Accuracy: 89.00%
Validation Accuracy: 84.91%



*   Hasil evaluasi menunjukkan bahwa model memiliki Training Accuracy sebesar 89.00% dan Validation Accuracy sebesar 84.91%.
*   Dengan ini. training accuracy (89%) menunjukkan bahwa model telah belajar dengan baik dari data training dan validation training (84.91%) menunjukkan bahwa model mampu generalisasi cukup baik pada data baru, meskipun sedikit lebih rendah dari akurasi training.

